content/blog/2023-11-28-context-switching-sprs-on-powerpc.md

```diff
@@ -0,0 +1,443 @@
+such as the LR value being used as the branch target address of the `blr`
+(branch to LR) instruction. Others might provide a way to view or interact with
+more fundamental CPU state, such as the Authority Mask Register (AMR) which the
+kernel uses to disable accesses to userspace memory while in supervisor mode.
+
+Most SPRs are per-CPU, much like GPRs. And, like GPRs, it makes sense for many
+of them to be tracked per-task, so that we can conceptually treat them as a
+per-task resource. But, depending on the particular SPR, writing to them can be
+_slow_. The highly used data-like SPRs such as LR, CTR (Count Register), etc.,
+are possible to [rename](https://en.wikipedia.org/wiki/Register_renaming),
+making them comparable to GPRs in terms of read/write performance. But others
+that affect the state of large parts of the CPU core, such as the [Data Stream
+Control Register (DSCR)](https://docs.kernel.org/next/powerpc/dscr.html), can
+take a while for the effects of changing them to be applied. Well, not so slow
+that you will notice the occasional access, but there's one case in the kernel
+that occurs extremely often and needs to change a lot of these SPRs to support
+our per-task abstraction: context switches.
+
+
+## Anatomy of a context switch
+
+Here we'll explore the actual function behind performing a context switch. We're
+interested in the SPR handling especially, because that's going to inform how we
+start tracking a new SPR on a per-task basis, so we'll be skimming over a lot of
+unrelated aspects.
+
+We start our investigation in the aptly named `context_switch()` function in
+`kernel/sched/core.c`.
+
+```c
+// kernel/sched/core.c
+
+/*
+ * context_switch - switch to the new MM and the new thread's register state.
+ */
+static __always_inline struct rq *
+context_switch(struct rq *rq, struct task_struct *prev,
+               struct task_struct *next, struct rq_flags *rf)
+```
+
+Along with some scheduling metadata, we see it takes a previous task and a next
+task. As we discussed above, the `struct task_struct` type describes a unit of
+execution that defines (among other things) how to set up the state of
+the CPU.
+
+This function starts off with some generic preparation and memory context
+changes, before getting to the meat of the function with
+`switch_to(prev, next,prev)`. This `switch_to()` call is actually a macro, which
+unwraps to a call to `__switch_to()`. It's also at this point that we enter the
+architecture specific implementation.
+
+> Aside: Changing the active memory mapping has no immediate effect on the
+> running code due to address space quadrants. In the hardware, the top two bits
+> of a 64 bit effective address determine what memory mapping is applied to
+> resolve it to a real address. If it is a userspace address (top two bits are
+> 0) then the configured mapping is used. But if it is a kernel address (top two
+> bits are 1) then the PID 0 mapping is always used. So our change to the memory
```

Comment:
"PID 0 mapping" - elaborate

Reply:
I added some extra words, but at the end of the day I'm writing this for people somewhat familiar with the concepts, but just hasn't looked into the details.